lib/reversetunnel/cache.go

```diff
@@ -0,0 +1,162 @@
+/*
+Copyright 2017 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reversetunnel
+
+import (
+	"net"
+	"sync"
+	"time"
+
+	"golang.org/x/crypto/ssh"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/auth/native"
+	"github.com/gravitational/teleport/lib/defaults"
+
+	"github.com/gravitational/trace"
+	"github.com/gravitational/ttlmap"
+)
+
+type certificateCache struct {
+	mu         sync.Mutex
+	cache      *ttlmap.TTLMap
+	authClient auth.ClientI
+}
+
+// NewHostCertificateCache creates a shared host certificate cache that is
+// used by the forwarding server.
+func NewHostCertificateCache(authClient auth.ClientI) (*certificateCache, error) {
+	cache, err := ttlmap.New(defaults.HostCertCacheSize)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &certificateCache{
+		cache:      cache,
+		authClient: authClient,
+	}, nil
+}
+
+// GetHostCertificate will fetch a certificate from the cache. If the certificate
+// is not in the cache, it will be generated, put in the cache, and returned.
+func (c *certificateCache) GetHostCertificate(addr string) (ssh.Signer, error) {
+	var certificate ssh.Signer
+	var err error
+	var ok bool
+
+	// extract the principal from the address
+	principal, _, err := net.SplitHostPort(addr)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	certificate, ok = c.get(principal)
```

Comment:
Isn't there a race condition here? Two callers come here at the same time, both see that there's no certificate for a certain address, then both proceed to generating it and then both save it and thus one will override another. I think this whole method should be under lock, no?

Reply:
Although placing everything under the lock might be a bad idea since "generateHostCert" makes API calls as far as I can see.

Reply:
it is not a race condition - it's just sometimes certs will get overwritten which is by design for exact the reason you mentioned - generating certs under lock is not scalable.

Reply:
@russjones can you add comments on why it's implemented this way.